src/3d/symbols/qgspointcloud3dsymbol_p.cpp

```diff
@@ -227,6 +267,179 @@ void QgsPointCloud3DSymbolHandler::makeEntity( Qt3DCore::QEntity *parent, const
+  bool hasParameterData = !outNormal.parameter.empty();
+
+  // first, get the points of the concerned node
+  std::vector<double> vertices( outNormal.positions.size() * 2 );
+  size_t idx = 0;
+  for ( int i = 0; i < outNormal.positions.size(); ++i )
+  {
+    vertices[idx++] = outNormal.positions.at( i ).x();
+    vertices[idx++] = outNormal.positions.at( i ).z();
+  }
+
+  // next, we also need all points of all parents nodes to make the triangulation (also external points)
+  IndexedPointCloudNode parentNode = n.parentNode();
+
+  int properPointsCount = outNormal.positions.count();
+  while ( parentNode.d() >= 0 )
+  {
+    processNode( pc, parentNode, context );
+    parentNode = parentNode.parentNode();
+  }
+
+  PointData filteredExtraPointData;
+
+  double span = pc->span();
+  double extraBoxFactor = 16 / span;
+  double extraX = extraBoxFactor * bbox.xExtent();
+  double extraZ = extraBoxFactor * bbox.zExtent();
+
+  // We keep all points in vertical direction to avoid odd triangulation if points are isolated on top
+  const QgsAABB extendedBBox( bbox.xMin - extraX, -std::numeric_limits<float>::max(), bbox.zMin - extraZ, bbox.xMax + extraX, std::numeric_limits<float>::max(), bbox.zMax + extraZ );
+
+  for ( int i = properPointsCount; i < outNormal.positions.count(); ++i )
+  {
+    const  QVector3D pos = outNormal.positions.at( i );
+    if ( extendedBBox.intersects( pos.x(), pos.y(), pos.z() ) )
+    {
+      filteredExtraPointData.positions.append( pos );
+      vertices.push_back( pos.x() );
+      vertices.push_back( pos.z() );
+      //vertSize += 2;
+
+      if ( hasColorData )
+        filteredExtraPointData.colors.append( outNormal.colors.at( i ) );
+      if ( hasParameterData )
+        filteredExtraPointData.parameter.append( outNormal.parameter.at( i ) );
+    }
+  }
+
+  outNormal.positions.resize( properPointsCount );
+  if ( hasColorData )
+    outNormal.colors.resize( properPointsCount );
+  if ( hasParameterData )
+    outNormal.parameter.resize( properPointsCount );
+
+  outNormal.positions.append( filteredExtraPointData.positions );
+  outNormal.colors.append( filteredExtraPointData.colors );
+  outNormal.parameter.append( filteredExtraPointData.parameter );
+
+  try
+  {
+    // Triangulation happens here
+    delaunator::Delaunator triangulation( vertices );
+
+    QVector<QVector3D> normals( outNormal.positions.count(), QVector3D( 0.0, 0.0, 0.0 ) );
+    const std::vector<size_t> &triangleIndexes = triangulation.triangles;
+    for ( size_t i = 0; i < triangleIndexes.size(); i += 3 )
+    {
+      QVector<QVector3D> triangleVertices( 3 );
+      for ( size_t j = 0; j < 3; ++j )
+      {
+        size_t vertIndex = triangleIndexes.at( i + j );
+        triangleVertices[j] = outNormal.positions.at( vertIndex );
+      }
+      //calculate normals
+      const QVector3D v1( triangleVertices.at( 1 ) - triangleVertices.at( 0 ) );
+      const  QVector3D v2( triangleVertices.at( 2 ) - triangleVertices.at( 0 ) );
```

Comment:
I would probably avoid creating a temporary three-value QVector many times and have three QVector3D variables - but I guess I am micro-optimizing here :-)